New Dashboard widgets for 2.10 ER1

josemigallas opened the pull request:
⚠️ This is not to be merged yet, we're using this branch as base.

**What this PR does / why we need it**:

We're removing the tabs in favor of two smaller widgets showing the top 5 last updated products and backends. Also removing the list of services from the Context Switcher and replacing it for a link to a Products and a Backend APIs index pages.

**Which issue(s) this PR fixes** 

[Address UI scalability issues in API Management Service](https://issues.redhat.com/browse/APPDUX-462)

**Verification steps** 

* Provider dashboard shows the 2 new widgets
* Products/Backend index pages can be visited:
  * Searchbar works
  * Pagination works

Martouta commented on app/decorators/application_decorator.rb:
```diff
@@ -15,10 +15,11 @@ def self.collection_decorator_class
   private
 
-  API_KEYS = %w[id name system_name].freeze
+  API_KEYS = %w[id name system_name links apps_count backends_count unread_alerts_count products_count].freeze
```
We spoke about this 😄 Improvement in process ⚠️  👍

Martouta commented on app/decorators/backend_api_decorator.rb:
```diff
@@ -18,4 +18,18 @@ def as_json(options = {})
+  end
+
+  def products_count
+    object.services.size
```
I don't know how this is done internally, but if it does the following, it is scary as hell 😱
1. backend_api left outer join backend_api_config left outer join services
2. count distinct services.id where backend_api.id is this one

instead, we should do this:
1. backend_api left outer join backend_api_config
2. count distinct backend_api_config.service_id where backend_api.id is this one

josemigallas replied:
What about calling `service.backend_api_configs` like it's done in `Account > Backends`?
https://github.com/3scale/porta/blob/ed7ad0d57ab2c8de7d4241afd5008f89de31fc59/app/controllers/api/backend_usages_controller.rb#L16

guicassolato replied:
I see what you're suggesting and indeed it would be an improvement:

```ruby
object = BackendApi.find(2)

object.services.count
# SELECT COUNT(*) FROM `services` INNER JOIN `backend_api_configs` ON `services`.`id` = `backend_api_configs`.`service_id` WHERE `backend_api_configs`.`backend_api_id` = 2

object.backend_api_configs.select(:service_id).distinct.count
# SELECT DISTINCT COUNT(DISTINCT `backend_api_configs`.`service_id`) FROM `backend_api_configs` WHERE `backend_api_configs`.`backend_api_id` = 2
```

However, I wonder if it's worth the effort. The proper way of optimising this would be by caching this count, but we don't have the time to do this now. Perhaps we can move on with this the way it is and consider another round of improvements later, with caching and all?

Martouta replied:
Yep, makes sense all of it! 💪  👍

Martouta commented on app/decorators/service_decorator.rb:
```diff
@@ -56,4 +56,30 @@ def as_json(options = {})
+  end
+
+  def backends_count
+    backend_apis.size
```
inverse situation to https://github.com/3scale/porta/pull/2194/files#r501110645

josemigallas replied:
Would `service.backend_api_configs` be valid here, like it's in `Account > Backends`?
https://github.com/3scale/porta/blob/ed7ad0d57ab2c8de7d4241afd5008f89de31fc59/app/controllers/api/backend_usages_controller.rb#L16

Martouta commented on app/controllers/provider/admin/dashboards_controller.rb:
```diff
@@ -14,6 +14,16 @@ def show
+
+    @widget_products = @services.order(updated_at: :desc)
+                                .decorate
+                                .take(5)
```
if we use `paginate` instead as with the other controllers, how many would there be by default? is it worth it? in that case, it would be easier to DRY this up.

Martouta replied:
https://github.com/3scale/porta/pull/2194/files#r501113662

josemigallas replied:
There won't be pagination in the dashboard, only the a list of 5 items.

Martouta commented on app/controllers/api/services_controller.rb:
```diff
@@ -16,6 +17,15 @@ class Api::ServicesController < Api::BaseController
 
+  def index
+    activate_menu :dashboard
+    @services = current_user.accessible_services
+                            .order(updated_at: :desc)
+                            .paginate(pagination_params)
+                            .decorate
+                            .to_json(only: %i[name updated_at id], methods: %i[links apps_count backends_count unread_alerts_count])
```
we should definitely DRY this 😄

josemigallas replied:
> TLDR: similar code but no repetition.

Well, I think DRY is not always the best solution: 2 equal lines of code don't necessarily have to be repeated.
In this case and IMO the scopes are different enough (different models, different controllers) so that DRY is not applicable.
Also, today we order by the same column (updated_at), tomorrow we don't know, maybe hits.
As for `paginate`, I wouldn't obscure it behind some decorator but rather leave it explicit for a quicker reading (plus, we'd need to pass the params).
Finally, keep in mind that each to_json passes its own fields and methods.

Martouta commented on app/controllers/provider/admin/backend_apis_controller.rb:
```diff
@@ -1,11 +1,22 @@
 
+  def index
+    activate_menu :dashboard
+    @backend_apis = current_account.backend_apis
+                                   .order(updated_at: :desc)
+                                   .paginate(pagination_params)
+                                   .decorate
+                                   .to_json(only: %i[name updated_at id private_url], methods: %i[links products_count])
```
let's DRY this 😄

Martouta replied:
https://github.com/3scale/porta/pull/2194/files#r501113662

Martouta commented on app/controllers/provider/admin/dashboards_controller.rb:
```diff
@@ -14,6 +14,16 @@ def show
+                                .decorate
+                                .take(5)
+                                .to_json(only: %i[name updated_at id], methods: %i[links])
+    @widget_backends = current_account.backend_apis.includes(:services)
+                                                    .order(updated_at: :desc)
+                                                    .decorate
+                                                    .take(5)
+                                                    .to_json(only: %i[name updated_at id], methods: %i[links])
```
https://github.com/3scale/porta/pull/2194/files#r501113662

Martouta commented on app/controllers/provider/admin/dashboards_controller.rb:
```diff
@@ -14,6 +14,16 @@ def show
     @messages_presenter = current_presenter
     @unread_messages_presenter = unread_messages_presenter
+
+    @widget_products = @services.order(updated_at: :desc)
+                                .decorate
+                                .take(5)
+                                .to_json(only: %i[name updated_at id], methods: %i[links])
```
https://github.com/3scale/porta/pull/2194/files#r501113662

Martouta commented on app/decorators/service_decorator.rb:
```diff
@@ -56,4 +56,30 @@ def as_json(options = {})
+  end
+
+  def apps_count
+    cinstances.size
```
maybe a `sum` of all the `contracts_count` field in the DB for all the `application_plans` of this service.
that would be more efficient.

Martouta replied:
similarly to https://github.com/3scale/porta/pull/2194/files#r504635406, next round for improvements 👍

Martouta commented on app/controllers/api/services_controller.rb:
```diff
@@ -17,11 +17,11 @@ class Api::ServicesController < Api::BaseController
-                            .paginate(pagination_params)
-                            .decorate
-                            .to_json(only: %i[name updated_at id system_name], methods: %i[links apps_count backends_count unread_alerts_count])
+    @raw_services = current_user.accessible_services
```
If you are not going to use this variable outside this method, better make it a local variable of the method rather than an instance variable of the class that can be used all over the controller, the views and the helpers.

```suggestion
    raw_services = current_user.accessible_services
```

Martouta replied:
I see that you use it for the counter... then better `@total_services_count = ...`
The less 'global variables' with less 'global info', the better.

Martouta replied:
For a following PR 👍

Martouta commented on app/controllers/api/services_controller.rb:
```diff
@@ -14,6 +15,15 @@ class Api::ServicesController < Api::BaseController
     actions.sublayout 'api/service'
   end
 
+  def index
```
I see many tips (in the logs) to improve the performance.

![image](https://user-images.githubusercontent.com/11318903/95969577-34ea4a00-0e0f-11eb-9d27-2a4bc9c167cb.png)

Martouta replied:
For a following PR 👍

Martouta commented on app/javascript/src/BackendApis/components/IndexPage.js:
```diff
@@ -0,0 +1,163 @@
+          <div className="pf-c-toolbar__content">
+            <ToolbarItem>
+              <InputGroup className="api-table-search">
+                <TextInput placeholder="Find a Backend" name="findBackend" id="findProduct" type="search" aria-label="Find a product" />
```
```suggestion
                <TextInput placeholder="Find a Backend" name="findBackend" id="findProduct" type="search" aria-label="Find a backend" />
```

🙈

thomasmaas commented:
Looking good! 

The only thing we didn't really think about is that the current product shows up as the title of the context switcher, which is a bit weird as individual products/backends are no longer part of it.
<img width="308" alt="Screenshot 2020-10-14 at 14 20 15" src="https://user-images.githubusercontent.com/54224/95988173-cc0fcb80-0e28-11eb-8068-7034ee3bf15c.png">

Martouta commented on app/decorators/application_decorator.rb:
```diff
@@ -12,4 +12,8 @@ def api_selector_api_link
   end
+
+  def updated_at
+    object.updated_at.to_s :long if object.updated_at?
```
Where is all this called/used from? 🤔

josemigallas replied:
This parses `updated_at` to a human friendly format, depending on the user's time zone when calling `to_json`.

Martouta commented on app/models/backend_api.rb:
```diff
@@ -1,9 +1,17 @@
   include SystemName
   include ProxyConfigAffectingChanges::ModelExtension
 
+  self.allowed_search_scopes = %i[query]
+
+  scope :by_query, ->(query) do
+    options = {ids_only: true, per_page: 1_000_000, star: true, ignore_scopes: true, with: { }}
+    where(id: search(ThinkingSphinx::Query.escape(query), options))
+  end
+
```
As I mentioned in the description of https://github.com/3scale/porta/pull/2204, we'll later have to polish all this and add tests 👍

Martouta commented on test/integration/api/services_controller_test.rb:
```diff
@@ -12,6 +12,11 @@ def setup
 
   attr_reader :provider, :service
 
+  test 'index' do
+    get admin_services_path
+    assert_response :success
+  end
+
```
later we will have to improve this test

Martouta commented on test/integration/provider/admin/backend_apis_controller_test.rb:
```diff
@@ -13,6 +13,11 @@ class Provider::Admin::BackendApisControllerTest < ActionDispatch::IntegrationTe
 
   attr_reader :provider
 
+  test '#index' do
+    get provider_admin_backend_apis_path
+    assert_response :success
+  end
```
later we will have to improve this test

Martouta reviewed (dismissed):
Great work! 👏 It can be polished but let's consider it for another round knowing that this feature is for today 😉

thomasmaas reviewed (dismissed):
approved based on click testing in preview. search wasn't working yet due to indexing.

cntlsn commented:
looking good, great work!
a couple of things I noticed while clicking around in preview:
* In the two widgets in the Dashboard, clicking on a Product or Backend name wouldn't lead you to the relative Product or Backend page
* In both the Product Dashboard widget and the Product page, when clicking on the 'Edit' item of the kebab menu the page reloads instead of leading you to the Edit Product page.

damianpm reviewed (changes requested):
There are two issues to fix:
1.- Icons are not equal in Dashboard
![icons](https://user-images.githubusercontent.com/13486237/95992766-9bcb2b80-0e2e-11eb-8e8b-2af0970cf4d6.png)
2.- There is an error in `isPlainButtonAction`, this may be the cause you can't click on API names:
<img width="796" alt="Screenshot 2020-10-14 at 15 05 51" src="https://user-images.githubusercontent.com/13486237/95992959-cfa65100-0e2e-11eb-8163-8aa03cd7d9b7.png">

josemigallas commented:
> There are two issues to fix:
> 1.- Icons are not equal in Dashboard
> ![icons](https://user-images.githubusercontent.com/13486237/95992766-9bcb2b80-0e2e-11eb-8e8b-2af0970cf4d6.png)
> 2.- There is an error in `isPlainButtonAction`, this may be the cause you can't click on API names:
> <img alt="Screenshot 2020-10-14 at 15 05 51" width="796" src="https://user-images.githubusercontent.com/13486237/95992959-cfa65100-0e2e-11eb-8163-8aa03cd7d9b7.png">

1. Good catch, fixed 👍 
2. I added a comment there, it's supposed to be valid according to PF documentation.. but it might has been removed recently. It shouldn't be a problem, we can take a look at it afterwards.

thomasmaas commented:
@josemigallas I'd rather fix the broken links before merging, so we can deploy master to saas if we have to.

damianpm commented on app/javascript/src/Dashboard/components/APIDataListItem.jsx:
```diff
@@ -0,0 +1,90 @@
+          aria-labelledby="multi-actions-item1 multi-actions-action1"
+          id="actions-menu"
+          aria-label="Actions"
+          // TODO: throwing a warning, does this prop really exist?
+          isPlainButtonAction
```
I think we can simply remove this prop. According to PF docs is not required and I don't see any difference in the rendered output
![pf-docs](https://user-images.githubusercontent.com/13486237/95998969-b7860000-0e35-11eb-8de2-cb4221783d7a.png)
```suggestion
```

josemigallas replied:
OK sounds good to me!

damianpm reviewed (dismissed):
LGTM, please run tests when ready

thomasmaas reviewed (approved):
All mentioned issues looking good now in preview

Martouta commented on features/step_definitions/applications/applications_steps.rb:
```diff
@@ -168,7 +168,7 @@
 Given(/^I'm on that application page$/) do
   click_on 'Dashboard'
-  click_on 'Overview'
+  click_on 'default' # this is supposed to be the name of the service
```
```suggestion
  click_on @application.service.name
```

Looks more dynamic 🤔 

No need to be today 👍

Martouta commented on features/old/stats/provider_side.feature:
```diff
@@ -15,17 +15,20 @@ Feature: Provider stats
 
   Scenario: Stats access
     When I log in as provider "foo.example.com"
+    And I follow "default"
```
Ugly but let's polish later

Martouta reviewed (approved):
Great job! 👏

Martouta commented on features/old/services/switch.feature:
```diff
@@ -11,18 +12,19 @@ Feature: Services switch
     Given current domain is the admin domain of provider "foo.example.com"
     When I log in as provider "foo.example.com"
 
+  # FIXME!!
```
This should be the `@wip` instead of the next one 🙈 My bad

damianpm reviewed (approved):
🥳
